### ISSUE TITLE
Add a threaded mode

### ISSUE THREAD

**Issue by dertuxmalwieder** [rafalp/Misago#539]
In these early stages of development, a threaded mode (like reddit and Slashdot) could be easy to implement. Misago would have another advantage then, as MyBB will drop this feature with 2.0.

**Comment by rafalp** [comment]
Please note that „early stages” is only valid for ember.js frontend. Backend for forums is done all the way to posting and lacks only attachments, reports, polls and e-mail notifications.

As for threaded discussions, I was considering them more than once, but ultimately found them fit for loosely moderated comments, (like on Reddit or Hackernews), but nightmare for gathering knowledge, where ultimately few bits tha are valuable (or funny) will be brought to the top, with everything else floowed by race for best joke/pun, and in places where moderation is tought, discussions having similiar acitivity to  toughtly moderated forums.

**Comment by dertuxmalwieder** [comment]
Would it be too much work to just add a flag "this posting is a reply to ID xxxx"? Problem is that in long(ish) discussions there are quite a few topics standing side by side, so a posting on page 57 could be a direct reply to something on page 32 which would (of course) break the logic. "Loosely moderated comments" could also be exhaustingly long explanations about a still unfinished topic.

**Comment by rafalp** [comment]
I wish performant implementation would be as simple as having relation to "parent" post. The Reddit for example has implementation splitting comments graph into branches that are tracked and cached separately.

Relation to parent will most likely land in anyway, seeing how we are using Ember.js for UI, adding rollovers for quick preview of parent and direct replies, but I don't plan to go further than that. :]

**Comment by dertuxmalwieder** [comment]
You won't need to copy reddit's full feature set. I thought a Threaded Mode is just a different display mode of the same information (well, even less information since you only need to show 1 full posting at the same time)...?

**Comment by rafalp** [comment]
The "different display mode" is very loaded thing to say when you consider having to load 2d against 3d list from relational database so say, display discussion's tree. :]

**Comment by dertuxmalwieder** [comment]
Why don't you load the 3d list every time then?

**Comment by rafalp** [comment]
You can't do that in RDBMS without a) making posting costlier (and slower) or b) making heavy use of caching (or both). You could do that in one of those memory document stores that let you nest documents within documents but thats not the case here (and will never be).

**Comment by dertuxmalwieder** [comment]
Ah, I see. But, as you theoretically build the "whole thread tree" anyway when displaying a thread, why would you have to change the dimension at all? I mean AFAICS you know everything that's required for "threading" when you open a thread...

**Comment by rafalp** [comment]
Yes and no. You can't just pull all posts from thread into memory and then build graph from them because thats not going to work for forums running with smaller memory limits or with high traffic. You'll got to make cut-off somewhere.

For 2d list thats pretty easy. Just cut it into slices (pages) and render selected pages to users. That's easy and cheap to be done in DB.

But how you'll do it with tree? Reddit splits trees into subtrees that are cached aggressively and separately, and each subtree has limit like x topnodes and y subnodes for each topnode and z sub-sub nodes for each subnode so memory and computation usage is predictable and constant, but that makes implementation tricky and dependant on presence of fast and large memory-based cache like redis or memcached.

So to sum this up, best that can be expected to happen would Discourse's approach where posts display buttons that let you see to whom is this post replying, as well as to reach quickly to replies to post in question.

**Comment by dertuxmalwieder** [comment]
> best that can be expected to happen would Discourse's approach where posts display buttons that let you see to whom is this post replying, as well as to reach quickly to replies to post in question.

I'd vote for that. :-) Can I help?

**Comment by rafalp** [comment]
Sadly its still weeks if not months before I'll feel confident enough with implementation of posting in Ember.js UI before I'll consider groundwork laid for other commiter to step up with pull requests for that.

So no, not yet, and I can't say when time will be right for that :|

**Comment by rafalp** [comment]
New task for tracking first pass on this feature: #1542
